# Avalible Plugins for controller in ROS2 Foxy
# - DWB Controller (dwb_core::DWBLocalPlanner)
# - TEB Controller (teb_local_planner::TebLocalPlannerROS)
# - Regulated Pure Pursuit (nav2_regulated_pure_pursuit_controller::RegulatedPurePursuitController)
# 

controller_server:
  ros__parameters:
    use_sim_time: False
    controller_frequency: 20.0
    min_x_velocity_threshold: 0.001
    min_y_velocity_threshold: 0.2
    min_theta_velocity_threshold: 0.001
    failure_tolerance: 0.3
    progress_checker_plugin: "progress_checker"
    goal_checker_plugins: ["general_goal_checker"] 
    controller_plugins: ["FollowPath"]

    # Progress checker parameters
    progress_checker:
      plugin: "nav2_controller::SimpleProgressChecker"
      required_movement_radius: 0.2
      movement_time_allowance: 10.0
    # Goal checker parameters
    general_goal_checker:
      plugin: "nav2_controller::SimpleGoalChecker"
      xy_goal_tolerance: 0.20
      yaw_goal_tolerance: 0.20
      stateful: True
    # Regulated Pure-Pursuit parameters
    FollowPath:
      plugin: "nav2_regulated_pure_pursuit_controller::RegulatedPurePursuitController"
      desired_linear_vel: 1.0
      max_linear_accel: 4.0
      max_linear_decel: 4.0
      transform_tolerance: 0.1
      lookahead_dist: 0.6
      min_lookahead_dist: 0.2
      max_lookahead_dist: 0.9
      lookahead_time: 0.25
      
      use_rotate_to_heading: false
      rotate_to_heading_min_angle: 0.785
      rotate_to_heading_angular_vel: 1.8
      max_angular_accel: 3.2
      
      use_velocity_scaled_lookahead_dist: true
      min_approach_linear_velocity: 0.15
      approach_velocity_scaling_dist: 0.6

      use_collision_detection: true
      max_allowed_time_to_collision_up_to_carrot: 1.0
      use_fixed_curvature_lookahead: false
      curvature_lookahead_dist: 0.25

      use_regulated_linear_velocity_scaling: false
      use_cost_regulated_linear_velocity_scaling: false
      regulated_linear_scaling_min_radius: 0.9
      regulated_linear_scaling_min_speed: 0.25

      allow_reversing: true
      max_robot_pose_search_dist: 10.0
    
local_costmap:
  local_costmap:
    ros__parameters:
      update_frequency: 5.0
      publish_frequency: 2.0
      global_frame: odom
      robot_base_frame: ego_racecar/base_footprint
      use_sim_time: False
      rolling_window: true
      static_map: false
      width: 3
      height: 3
      resolution: 0.05
      robot_radius: 0.25
      plugins: ["stvl_layer", "inflation_layer"] # exclude "voxel_layer", costmap plugin are ordered!!!

      inflation_layer:
        plugin: "nav2_costmap_2d::InflationLayer"
        cost_scaling_factor: 3.0
        inflation_radius: 0.02

      # voxel_layer:
      #   plugin: "nav2_costmap_2d::VoxelLayer"
      #   enabled: True
      #   publish_voxel_map: True
      #   origin_z: 0.0
      #   z_resolution: 0.05
      #   z_voxels: 10
      #   max_obstacle_height: 2.0
      #   mark_threshold: 0
      #   observation_sources: pointcloud
      #   pointcloud: # no frame set, uses frame from message
      #     topic: /livox/lidar
      #     max_obstacle_height: 2.0
      #     min_obstacle_height: 0.22
      #     obstacle_max_range: 0.5
      #     obstacle_min_range: 0.0
      #     raytrace_max_range: 3.0
      #     raytrace_min_range: 0.0
      #     clearing: True
      #     marking: True
      #     data_type: "PointCloud2"
      
      stvl_layer:
        plugin: "spatio_temporal_voxel_layer/SpatioTemporalVoxelLayer"
        enabled: true
        voxel_decay: 1.0
        decay_model: 0
        voxel_size: 0.05
        track_unknown_space: true
        unknown_threshold: 15
        mark_threshold: 0
        update_footprint_enabled: true
        combination_method: 1
        origin_z: 0.0
        publish_voxel_map: true
        transform_tolerance: 0.2
        mapping_mode: false
        map_save_duration: 60.0
        observation_sources: pointcloud
        pointcloud:
          data_type: PointCloud2
          topic: /livox/lidar
          marking: true
          clearing: true
          obstacle_range: 3.0
          min_obstacle_height: 0.22
          max_obstacle_height: 2.0
          expected_update_rate: 0.0
          observation_persistence: 0.0
          inf_is_valid: false
          filter: "voxel"
          voxel_min_points: 0
          clear_after_reading: true
          max_z: 7.0
          min_z: 0.1
          vertical_fov_angle: 0.8745
          horizontal_fov_angle: 1.048
          decay_acceleration: 15.0
          model_type: 1

      # static_layer:
      #   map_subscribe_transient_local: True
      always_send_full_costmap: True
        